tests/unit/core.js

```diff
@@ -1952,3 +1952,139 @@ exports.duplicateProto = function (test) {
+    "}"
+  ];
+
+  TestRun(test, "must be called")
+    .addError(3, "You must call 'super' in derived constructors.")
+    .test(code, { esnext: true });
+
+  var code2 = [
+    "super();"
+  ];
+
+  TestRun(test, "can't be used in global scope")
+    .addError(1, "'super' can only be called in derived constructors.")
+    .test(code2, { esnext: true });
+
+  var code3 = [
+    "function x() {",
+    "  super();",
+    "}",
+    "class A {}",
+    "class B extends A {",
+    "  constructor() {",
+    "    (function() {",
+    "      super();",
+    "    })();",
+    "  }",
+    "}",
+    "function y(param = super()) {}"
+  ];
+
+  TestRun(test, "can't be used in functions")
+    .addError(2, "'super' can only be called in derived constructors.")
+    .addError(6, "You must call 'super' in derived constructors.")
+    .addError(8, "'super' can only be called in derived constructors.")
+    .addError(12, "'super' can only be called in derived constructors.")
+    .test(code3, { esnext: true });
+
+  var code4 = [
+    "var x = {",
+    "  constructor() {",
+    "    super();",
+    "  }",
+    "};"
+  ];
+
+  TestRun(test, "can't be used in object methods")
+    .addError(3, "'super' can only be called in derived constructors.")
+    .test(code4, { esnext: true });
+
+  var code5 = [
+    "class A {",
+    "  constructor() {",
+    "    super();",
+    "  }",
+    "}",
+    "class B extends null {",
+    "  constructor() {",
+    "    super();",
+    "  }",
+    "}"
+  ];
+
+  TestRun(test, "can't be used in classes which don't extend another class")
+    .addError(3, "'super' can only be called in derived constructors.")
+    .addError(8, "'super' can only be called in derived constructors.")
+    .test(code5, { esnext: true });
+
+  var code6 = [
+    "class A {}",
+    "class B extends A {",
+    "  static constructor() {",
+    "    super();",
+    "  }",
+    "  method() {",
+    "    super();",
+    "  }",
+    "}"
+  ];
+
+  TestRun(test, "can't be used in class methods other than `constructor`")
+    .addError(4, "'super' can only be called in derived constructors.")
+    .addError(7, "'super' can only be called in derived constructors.")
+    .test(code6, { esnext: true });
+
+  var code7 = [
+    "class A {}",
+    "class B extends A {",
+    "  constructor() {",
+    "    (() => super())();",
+    "  }",
+    "}",
+    "(() => super())();"
+  ];
+
+  TestRun(test, "can be used in arrow functions in `constructor`")
+    .addError(7, "'super' can only be called in derived constructors.")
+    .test(code7, { esnext: true });
+
+  test.done();
+};
+
+exports["this must be used after super()"] = function (test) {
+  var code = [
+    "class A {}",
+    "class B extends A {",
+    "  constructor() {",
+    "    super();",
+    "    this.x = 1;",
+    "  }",
+    "}"
+  ];
+
+  TestRun(test).test(code, { esnext: true });
+
+  var code2 = [
+    "class A {}",
+    "class B extends A {",
+    "  constructor() {",
+    "    this.x = 1;",
```

Comment:
Sorry, I didn't mean to make it warn.. I'm not sure whether it should or shouldn't..

Presumably in this case
```
constructor() {
   var a = ()=>this.x = 1;
   super();
   a();
}
```

it works?

in this case..
```
(()=>a)();
let a;
```

we don't error about use in the TDZ because we don't know that the function is immediately invoked or not - we do however give a latdef warning, if that option is on. There is no such latdef for super type option (it seems to much of a corner case to create one), so we have to decide.

@jugglinmike do you think we should warn in this case or not?